luigi/execution_summary.py

```diff
@@ -0,0 +1,370 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2015-2015 Spotify AB
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+This module provide the function :py:func:`summary` that is used for printing
+an execution summary at the end of luigi invocations. For example:
+
+    .. code:: console
+
+            $ luigi --module examples.execution_summary_example examples.EntryPoint --local-scheduler
+            ...
+            ... lots of spammy output
+            ...
+            INFO: There are 11 pending tasks unique to this worker
+            INFO: Worker Worker(salt=843361665, workers=1, host=arash-spotify-T440s, username=arash, pid=18534) was stopped. Shutting down Keep-Alive thread
+            INFO:
+           ===== Luigi Execution Summary =====
+
+            Scheduled 210 tasks of which:
+            * 195 were already done:
+                - 195 examples.Bar(num=5...199)
+            * 1 ran successfully:
+                - 1 examples.Boom(num=0)
+            * 14 were left pending (following subcategories are not necessarily distinct):
+                * 1 were external dependencies:
+                    - 1 MyExternal()
+                * 13 had missing dependencies:
+                    - 1 examples.EntryPoint()
+                    - 10 examples.Foo(num=100, num2=0...9)
+                    - examples.DateTask(date=1998-03-23, num=1) and examples.DateTask(date=1998-03-23, num=0)
+
+            This progress looks :| because there were missing tasks
+
+            ===== Luigi Execution Summary =====
+"""
+
+import textwrap
+import datetime
+
+
+def _partition_tasks(worker):
+    """
+    Takes a worker and sorts out which tasks were completed, already done, failed and still pending.
+    Still_pending_not_ext is only used to get upstream_failure, upstream_missing_dependency and run_by_other_worker
+    """
+    task_history = worker._add_task_history
+    pending_tasks = {task for(task, status, ext) in task_history if status == 'PENDING'}
+    set_tasks = {}
```

Comment:
ok that's a good reason :) might be worth taking a look at it either way, classes are useful.

functional purists would probably complain about using a class here, but i'm pragmatic and think it would lead to somewhat more simple code. you could potentially make `partition_tasks` a constructor for a class and then make everything else class methods